### PR TITLE
Add MongoDB backend

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

MongoDB is a classical database for LRSs such as LearningLocker or Apereo Learning Analytics Initiative. Supporting this backend seems a must-have for Ralph.

## Proposal

- [x] add LearningLocker parser
- [x] add MongoDB backend
- [x] ~add learning locker parser~ **edit: postponed to a new PR**

### CODE REVIEW

File: src/ralph/backends/database/mongo.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        """Converts `stream` lines to Mongo documents."""
+
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+            document = {
+                "_id": ObjectId(
+                    hashlib.sha256(
+                        bytes(statement.get("id", None), "utf-8")
+                    ).hexdigest()[:24]
+                ),
```
Comment (jmaupetit) [inline src/ralph/backends/database/mongo.py:66]: This section requires at least a comment to explain why we need to enforce the ID.

File: src/ralph/backends/database/mongo.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        for document in self.to_documents(stream):
+
+            batch.append(document)
+            if len(batch) < chunk_size:
+                continue
+
+            success += self.bulk_import(batch, ignore_errors=ignore_errors)
+            batch = []
+
+        # Edge case: if the total number of documents is lower that the chunk size
```
Comment (jmaupetit) [inline src/ralph/backends/database/mongo.py:114]:
```suggestion
        # Edge case: if the total number of documents is lower than the chunk size
```

**Review by jmaupetit — COMMENTED** [review]
Ooops, LearningLocker parser tests are missing. Will do this asap.

**Comment by jmaupetit** [comment]
That said, `ElasticSearchParser` are also missing @SergioSim :wink:

**Comment by SergioSim** [comment]
> That said, `ElasticSearchParser` are also missing @SergioSim wink

The parser was covered in the [cli test](https://github.com/openfun/ralph/blob/master/tests/test_cli.py#L148) but I agree that it is always a good idea to test it separately) Well spotted) I will add the test)

File: src/ralph/backends/database/mongo.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    name = "mongo"
+
+    def __init__(
+        self,
+        connection_uri: str = MONGO_URI,
+        database: str = MONGO_DATABASE,
+        collection: str = MONGO_COLLECTION,
+        client_options: dict = None,
+    ):
+        """Instanciates the Mongo client.
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:31]: Once day, I'll remember this. One day. :grimacing:

File: src/ralph/backends/database/mongo.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        except BulkWriteError as error:
+            if not ignore_errors:
+                raise error
+            logger.warning(
+                "Bulk importation failed for current documents chunk but you choose "
+                "to ignore it.",
+            )
+            return 0
+
+        success = len(new_documents.inserted_ids)
```
Comment (quitterie-lcs) [inline src/ralph/backends/database/mongo.py:91]: Shouldn't the name be more explicit?

File: src/ralph/backends/database/mongo.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+            document = {
+                "_id": ObjectId(
+                    hashlib.sha256(
+                        bytes(statement.get("id", None), "utf-8")
+                    ).hexdigest()[:24]
+                ),
+                "statement": statement,
```
Comment (SergioSim) [inline src/ralph/backends/database/mongo.py:72]: Why do we put the `statement` inside a "statement" key? Could we write the input data directly at the root level?

File: src/ralph/backends/database/mongo.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            }
+
+            yield document
+
+    def bulk_import(self, batch: list, ignore_errors: bool = False):
+        """Bulk import a batch of documents to selected database collection."""
```
Comment (SergioSim) [inline src/ralph/backends/database/mongo.py:78]:
```suggestion
        """Inserts a batch of documents into the selected database collection."""
```

**Review by SergioSim — COMMENTED** [review]
This looks awesome! Have only a few minor suggestions & questions

File: src/ralph/backends/database/mongo.py (hunk 5 above)
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:72]: I mimicked learning locker's schema: https://docs.learninglocker.net/http-statements/

File: src/ralph/backends/database/mongo.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        to compute a unique MongoDB Object ID. This ensures that we will not
+        duplicate statements in our database.
+        """
+
+        for line in stream:
+            statement = json.loads(line) if isinstance(line, str) else line
+            document = {
+                "_id": ObjectId(
+                    hashlib.sha256(
+                        bytes(statement.get("id", None), "utf-8")
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:69]: I think it's not the backend's responsibility to create such identifier for a statement that has been externally generated. I would prefer considering that the LRS has generated the statement identifier. But you are right: we should raise an error if the statement does not container an identifier at this stage.

File: .env.dist [hunk 8]
```diff
@@ -40,6 +40,12 @@ RALPH_APP_DIR=/app/.ralph
 # RALPH_ES_TEST_HOSTS=http://elasticsearch:9200
 # RALPH_ES_TEST_INDEX=test-index
 
+# MONGO database backend
+
+# RALPH_MONGO_COLLECTION=marsha
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .env.dist:45]: This is a commented example, but we can choose a more relevant / generic name. Do you have a suggestion?

File: src/ralph/backends/database/mongo.py (hunk 4 above)
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:91]: `success`?

File: Makefile [hunk 9]
```diff
@@ -108,12 +108,14 @@ logs: ## display app logs (follow mode)
 
 run: ## alias for run-es
```
[the remark this answers is not in the document]
Comment (quitterie-lcs) [reply Makefile:109]: Shouldn't we delete this one and just keep the `run-es` and `run-mongo` commands ?

File: .env.dist [hunk 10]
```diff
@@ -40,6 +40,12 @@ RALPH_APP_DIR=/app/.ralph
...
+# RALPH_MONGO_DATABASE=statements
+# RALPH_MONGO_URI=mongodb://localhost:27017/
+
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .env.dist:48]: I don't get it. Do you suggest to suppress the empty line before `RALPH_MONGO` vars definition?

[the remark this answers is not in the document]
Comment (SergioSim) [reply .env.dist:48]: It's a suggestion to add additional "TEST" environment variables which are used in the tests)

**Comment by jmaupetit** [comment]
I've postponed the learning locker events parser to a new PR as this work is starting to grow and would preferably be reviewed separately.

File: src/ralph/backends/database/mongo.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (SergioSim) [reply src/ralph/backends/database/mongo.py:69]: Sounds good) Could we adapt the error message to be more explicit about what went wrong? Maybe wrap it in a try-catch or access the id field directly (`statement["id"]`) to get a `KeyError: 'id'` in that case? WDYT?

File: .env.dist (hunk 10 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .env.dist:48]: I think I need more coffee :sweat_smile:

File: .env.dist (hunk 8 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply .env.dist:45]: Deal!

File: src/ralph/backends/database/mongo.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        self.database = getattr(self.client, database)
+        self.collection = getattr(self.database, collection)
+
+    def get(self, chunk_size: int = 500):
+        """Gets collection documents and yields them."""
+
+        for document in self.collection.find(batch_size=chunk_size):
+            # Make the document json-serializable
+            document.update({"_id": str(document.get("_id"))})
+            yield document
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:53]: > This is true when the xAPI statement is stored at the root of the document, however, LearningLocker, as you mentioned, stores it inside a "statement" field which has next to it a "person" field containing another MongoDB _id field.

I think you are making a point that needs to be discussed furthermore about the "schema" we use to store statements in the MongoDB backend. Let's do this IRL!

> Alternatively, we could select only the "statement" field in the find query to only work with xAPI data in the case of LearningLocker, however, this would make the Mongo backend limited to LRSes following the LearningLocker schema.

I think we should stay agnostic. The `learninglocker` parser is responsible to extract statements from a leaninglocker Mongo database.

File: src/ralph/backends/database/mongo.py (hunk 4 above)
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:91]: Deal (`inserted_count`).

File: src/ralph/backends/database/mongo.py (hunk 5 above)
Comment (SergioSim) [reply src/ralph/backends/database/mongo.py:72]: Could the MongoDB backend be more generic in this regard? For example, [lxHive LRS](https://github.com/g3i/lxHive) seems to store xAPI statements at the root level of the MongoDB document.

File: src/ralph/backends/database/mongo.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:69]: Done.

File: src/ralph/backends/database/mongo.py (hunk 5 above)
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:79]: That's interesting! We should had this to our future IRL discussion.

File: src/ralph/backends/database/mongo.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply src/ralph/backends/database/mongo.py:54]: As mention in another comment, if we allow custom queries to get documents in database backends, it is no longer required to write dedicated parsers to support various LRS's "schema".

Concerning binary data, @SergioSim will declare an issue to keep in mind considering `python-bsonjs` if we encounter such use-case.

**Comment by jmaupetit** [comment]
All comments have been addressed now. I would be pleased to have a new review to merge this work @quitterie-lcs & @SergioSim :pray:

File: tests/backends/database/test_mongo.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+    assert hasattr(backend.client, MONGO_TEST_DATABASE)
+    database = getattr(backend.client, MONGO_TEST_DATABASE)
+    assert hasattr(database, MONGO_TEST_COLLECTION)
+
+
+def test_backends_database_mongo_get_method(mongo):
+    """Test the mongo backend get method."""
+
+    # Create records
+    documents = MongoDatabase.to_documents([{"id": "foo"}, {"id": "bar"}])
```
Comment (quitterie-lcs) [inline tests/backends/database/test_mongo.py:39]: What do you think of using the `fixed_dictionaries` strategy from hypothesis to generate statements?

Comment (jmaupetit) [reply tests/backends/database/test_mongo.py:39]: Maybe it's too convoluted for those tests? Also it might be less efficient.

Comment (quitterie-lcs) [reply tests/backends/database/test_mongo.py:39]: you're right, let's keep the KISS principle in mind

File: src/ralph/backends/database/mongo.py [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        try:
+            new_documents = self.collection.insert_many(batch)
+        except BulkWriteError as error:
+            if not ignore_errors:
+                raise error
+            logger.warning(
+                "Bulk importation failed for current documents chunk but you choose "
+                "to ignore it.",
+            )
+            return 0
```
Comment (SergioSim) [inline src/ralph/backends/database/mongo.py:96]:
```suggestion
            return error.details["nInserted"]
```
It seems pymongo inserts any documents prior to the first failure and omits all subsequent insertions.

**Review by SergioSim — APPROVED** [review]
Looks good to me) Have only one minor suggestion)

Image

File: src/ralph/backends/database/mongo.py (hunk 13 above)
Comment (SergioSim) [reply src/ralph/backends/database/mongo.py:96]: Apropos, could it be helpful to include the number of inserted/skipped documents in the warning log message? WDYT?
